### PR TITLE
Asynchronous client using rlp_03 provided methods

### PR DESCRIPTION
**Pull request by kortemik**

### CODE REVIEW

File: src/main/java/com/teragrep/rlp_03/ConnectContext.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                    socketFactory.create(socketChannel),
+                    interestOps,
+                    frameDelegate
+            );
+            // change attachment to established -> ConnectionContext
+            selectionKey.attach(connectionContext);
+
+            interestOps.add(SelectionKey.OP_READ);
+
+            LOGGER.info("ready");
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/ConnectContext.java:121]: Sounds spammy and not informative

File: src/main/java/com/teragrep/rlp_03/EventLoop.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        LOGGER.debug("readyKeys <{}>", readyKeys);
+
+        Set<SelectionKey> selectionKeys = selector.selectedKeys();
+        LOGGER.debug("selectionKeys <{}> ", selectionKeys);
+        for (SelectionKey selectionKey : selectionKeys) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER
+                        .debug(
+                                "selectionKey <{}>: " + "isValid <{}>, " + "isConnectable <{}>, "
+                                        + "isAcceptable <{}>, " + "isReadable <{}>, " + "isWritable <{}>",
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/EventLoop.java:85]: Why string concat

File: src/main/java/com/teragrep/rlp_03/EventLoop.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                        );
+            }
+
+            if (selectionKey.isAcceptable()) {
+                // ListenContext
+                ListenContext listenContext = (ListenContext) selectionKey.attachment();
+                listenContext.handleEvent(selectionKey);
+            }
+            else if (selectionKey.isConnectable()) {
+                if (selectionKey.isConnectable()) {
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/EventLoop.java:97]: Doubled same if?

File: src/main/java/com/teragrep/rlp_03/ServerFactory.java [hunk 4]
```diff
@@ -55,14 +55,22 @@
...
 import java.util.concurrent.TimeUnit;
 import java.util.function.Supplier;
 
-import static java.nio.channels.SelectionKey.OP_ACCEPT;
+/*
+ * TODO take in a "SocketPoll" instance or create new one in secondary constructor
+ *  TODO use this to get a selector, for creating a ListenInstance
+ *   TODO the runnable "Server" is not actually a ListenInstance but a "SocketPoll", perhaps we keep "Server" as a shortcut
+ *    TODO create a ClientFactory too?
+ *     TODO client is not a runnable but just yet another thing in the "SocketPoll" loop, just like a ListenInstance
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/ServerFactory.java:68]: Are these important for this PR or future stuff

File: src/main/java/com/teragrep/rlp_03/ServerFactory.java [hunk 5]
```diff
@@ -91,6 +99,7 @@ public ServerFactory(Config config, TLSConfig tlsConfig, Supplier<FrameDelegate>
                 TimeUnit.SECONDS,
                 new LinkedBlockingQueue<>()
         );
+        // FIXME             executorService.shutdown();
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/ServerFactory.java:102]: FIXME

File: src/main/java/com/teragrep/rlp_03/context/ConnectionContextImpl.java [hunk 6]
```diff
@@ -123,6 +123,7 @@ public void close() {
 
     @Override
     public void handleEvent(SelectionKey selectionKey) {
+        // TODO refactor this into Strategy pattern
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/ConnectionContextImpl.java:126]: Is this for this PR or future task

File: src/test/java/com/teragrep/rlp_03/ManualPerformanceTest.java [hunk 7]
```diff
@@ -149,11 +149,10 @@ public void run() {
...
                 long rate = (end - start) / (interval / 1000);
 
-                LOGGER
-                        .info(
-                                "Current records per second rate <{}>, threads <{}>, tasksQueue.size <{}>", rate,
-                                server.executorService.getActiveCount(), server.executorService.getQueue().size()
-                        );
+                LOGGER.info("Current records per second rate <{}>, threads <{}>, tasksQueue.size <{}>", rate//,
+                //server.executorService.getActiveCount(),
+                //server.executorService.getQueue().size()
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/ManualPerformanceTest.java:154]: Commented out code

File: src/test/java/com/teragrep/rlp_03/ServerShutdownTest.java [hunk 8]
```diff
@@ -64,10 +64,10 @@ public void testServerShutdownSingleThread() {
 
             server.startup.waitForCompletion();
 
-            Assertions.assertFalse(server.executorService.isShutdown());
+            //Assertions.assertFalse(server.executorService.isShutdown());
             server.stop();
             serverThread.join();
-            Assertions.assertTrue(server.executorService.isShutdown());
+            //Assertions.assertTrue(server.executorService.isShutdown());
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/ServerShutdownTest.java:70]: Disable test if theres something that breaks it, don't comment out assertions

File: src/test/java/com/teragrep/rlp_03/ServerShutdownTest.java [hunk 9]
```diff
@@ -84,10 +84,10 @@ public void testServerShutdownMultiThread() {
...
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/ServerShutdownTest.java:90]: Disable test if theres something that breaks it, don't comment out assertions

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+            Thread serverThread = new Thread(server);
+            serverThread.start();
+
+            server.startup.waitForCompletion();
+        });
+    }
+
+    @Test
+    public void t() throws IOException, InterruptedException {
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/client/ClientTest.java:98]: Not descriptive name

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+                LOGGER.debug("openResponse <[{}]>", openResponse);
+                String syslogResponse = syslog.get();
+                LOGGER.debug("syslogResponse <[{}]>", syslogResponse);
+                Assertions.assertEquals("200 OK", syslogResponse);
+                String closeResponse = close.get();
+                LOGGER.debug("closeResponse <[{}]>", closeResponse);
+            }
+            catch (ExecutionException executionException) {
+                throw new RuntimeException(executionException); // TODO
+            }
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/client/ClientTest.java:179]: Don't try/catch, use assertions instead

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+            try {
+                Thread.sleep(1000L);
+            }
+            catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/client/ClientTest.java:190]: Nested try/catch blocks makes verifying the tests hard, should not be used

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+                // TODO implement better handling for hint frames
+                if (txn == 0) {
+                    return true;
+                }
+
+                CompletableFuture<String> future = pendingTransactions.remove(txn);
+
+                if (future == null) {
+                    throw new IllegalStateException("txn not pending <[" + txn + "]>");
+                }
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/client/ClientTest.java:134]: Assert instead

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+        CompletableFuture<ConnectionContext> readyContextFuture = new CompletableFuture<>();
+
+        Consumer<ConnectionContext> connectionContextConsumer = connectionContext -> {
+            LOGGER.debug("connectionContext ready");
+            readyContextFuture.complete(connectionContext);
+        };
+
+        // TODO perhaps <Integer, Future<Something>> ?
+        // FIXME what should be Something, RelpFrame is immediately deallocated after FrameDelegate
+        // TODO design better: Futures are not optimal for multi-complete/disruptor pattern
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/client/ClientTest.java:114]: Are these for this PR or future tasks

**Review by StrongestNumber9 — COMMENTED** [review]
See comments

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java (hunk 13 above)
Comment (kortemik) [reply src/test/java/com/teragrep/rlp_03/client/ClientTest.java:134]: exception is part of the logic required, meaning it is not for testing. futures are tested elsewhere.

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java (hunk 14 above)
Comment (kortemik) [reply src/test/java/com/teragrep/rlp_03/client/ClientTest.java:114]: future tasks

File: src/main/java/com/teragrep/rlp_03/EventLoop.java (hunk 2 above)
Comment (kortemik) [reply src/main/java/com/teragrep/rlp_03/EventLoop.java]: left by auto-indentation, is now corrected

File: src/main/java/com/teragrep/rlp_03/EventLoop.java (hunk 3 above)
Comment (kortemik) [reply src/main/java/com/teragrep/rlp_03/EventLoop.java]: removed now, was left due to refactoring

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java (hunk 10 above)
Comment (kortemik) [reply src/test/java/com/teragrep/rlp_03/client/ClientTest.java]: renamed to testClientFunctionality

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java (hunk 11 above)
Comment (kortemik) [reply src/test/java/com/teragrep/rlp_03/client/ClientTest.java]: unwrapped, last is required because ExecutionException is a checked type

**Comment by kortemik** [comment]
@StrongestNumber9 please recheck

File: src/test/java/com/teragrep/rlp_03/client/ClientTest.java [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+            LOGGER.debug("syslogResponse <[{}]>", syslogResponse);
+            Assertions.assertEquals("200 OK", syslogResponse);
+            String closeResponse = close.get();
+            LOGGER.debug("closeResponse <[{}]>", closeResponse);
+            Assertions.assertEquals("", closeResponse);
+
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (StrongestNumber9) [inline src/test/java/com/teragrep/rlp_03/client/ClientTest.java:186]: I think this could also be wrapped in Assertions.assertAll bit, otherwise lgtm

**Review by StrongestNumber9 — APPROVED** [review]
Maybe the last catch in test could be changed, otherwise lgtm

**Comment by kortemik** [comment]
@StrongestNumber9 mind re-checking after latest changes?

**Pull request closed without merging**